Drop-In/src/main/java/com/braintreepayments/api/dropin/adapters/AvailablePaymentMethodNonceList.java

```diff
@@ -0,0 +1,59 @@
+package com.braintreepayments.api.dropin.adapters;
+
+import android.content.Context;
+
+import com.braintreepayments.api.dropin.DropInRequest;
+import com.braintreepayments.api.dropin.utils.PaymentMethodType;
+import com.braintreepayments.api.models.CardNonce;
+import com.braintreepayments.api.models.Configuration;
+import com.braintreepayments.api.models.GooglePaymentCardNonce;
+import com.braintreepayments.api.models.PayPalAccountNonce;
+import com.braintreepayments.api.models.PaymentMethodNonce;
+import com.braintreepayments.api.models.VenmoAccountNonce;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+class AvailablePaymentMethodNonceList {
+
+    final private List<PaymentMethodNonce> items;
+
+    AvailablePaymentMethodNonceList(Context context, Configuration configuration, List<PaymentMethodNonce> paymentMethodNonces, DropInRequest dropInRequest, boolean googlePayEnabled, boolean unionpaySupported) {
+        items = new ArrayList<>();
+
+        for (PaymentMethodNonce paymentMethodNonce: paymentMethodNonces) {
+            boolean shouldAddPaymentMethod = false;
+
+            if (paymentMethodNonce instanceof PayPalAccountNonce && dropInRequest.isPayPalEnabled() && configuration.isPayPalEnabled()) {
+                shouldAddPaymentMethod = true;
+            } else if (paymentMethodNonce instanceof VenmoAccountNonce && dropInRequest.isVenmoEnabled() && configuration.getPayWithVenmo().isEnabled(context)) {
+                shouldAddPaymentMethod = true;
+            } else if (paymentMethodNonce instanceof CardNonce && dropInRequest.isCardEnabled()) {
+                Set<String> supportedCardTypes =
+                        new HashSet<>(configuration.getCardConfiguration().getSupportedCardTypes());
+                if (!unionpaySupported) {
+                    supportedCardTypes.remove(PaymentMethodType.UNIONPAY.getCanonicalName());
+                }
+                if (supportedCardTypes.size() > 0) {
+                    shouldAddPaymentMethod = true;
+                }
```

Comment:
Where did this logic around union pay come from? It seems a little odd, because as long as the merchant supports union pay **and** another card type, a vaulted union pay card could be displayed even if `unionpaySupported` is false.

Reply:
Yeah I agree it's a direct port of the [SupportedPaymentMethodsAdapter](https://github.com/braintree/braintree-android-drop-in/blob/master/Drop-In/src/main/java/com/braintreepayments/api/dropin/adapters/SupportedPaymentMethodsAdapter.java) logic. We put it in a data structure to decouple from Android. There wasn't an easy way to extract and reuse the functionality since one filters enums and the other filters nonce instances.

Reply:
I'm having trouble wrapping my head around the logic in `SupportedPaymentMethodsAdapter`, but it seems like we shouldn't display union pay cards if union pay isn't supported. Is there a way to check if the nonce is a union pay nonce, and then only display it if union pay is supported?

For other cards, it seems like we should always display them as long as cards are enabled on `DropInRequest`. I suppose there's an edge case where a customer vaults a card, and then the merchant turns off support for that card in the Control Panel, but I don't think we should be handling that on the front-end.

Reply:
Yeah this is a good call I don't think there's a true equivalent for that if-else clause in this data structure. Supported payment methods adapter is basically trying to determine if it should display `PaymentMethodType.UNKNOWN` which is basically the `Add Card` button. This would guard against Union Pay nonces being displayed iff Union Pay was a merchant's only supported card brand. 🤔

Reply:
And I think we would have to use `UnionPay.fetchCapabilities()` to check if a Card Nonce is a Union Pay Nonce 😞

Reply:
Yeah this is tough I think we would need a single network call per nonce to check if UnionPay I'm unaware of any method to check if it's UnionPay without making the round trip.

Reply:
We ended up simplifying this to only check if cards are enabled in the request and any supported card type exists. We can revisit UnionPay in the future if needed.